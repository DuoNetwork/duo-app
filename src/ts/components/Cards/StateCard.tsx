import { Tooltip } from 'antd';
import * as d3 from 'd3';
import successIcon from 'images/stauts/success.svg';
import warningIcon from 'images/stauts/warning.svg';
import moment from 'moment';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import { IDualClassStates } from 'ts/common/types';
import { getDualClassAddressByTypeTenor } from 'ts/common/wrappers';
import { SDivFlexCenter } from '../_styled';
import { SCard, SCardExtraDivSolid, SCardList, SCardListProgressBar, SCardTitle } from './_styled';

interface IProps {
	type: string;
	tenor: string;
	locale: string;
	states: IDualClassStates;
	mobile?: boolean;
}

export default class StateCard extends React.Component<IProps> {
	public render() {
		const { states, mobile, locale, tenor, type } = this.props;
		const tooltioText =
			states.state === CST.CTD_TRADING
				? CST.TT_TRADING_STATE[locale]
				: states.state === CST.CTD_INCEPTION
				? 'Inception state, please wait until the system is started.'
				: CST.TT_RESET_STATE[locale];
		return (
			<SCard
				title={
					<SCardTitle>
						<a
							className="tag-content"
							href={
								'https://' +
								(__KOVAN__ ? 'kovan.' : '') +
								'etherscan.io/address/' +
								getDualClassAddressByTypeTenor(type, tenor).custodian.address
							}
							target="_blank"
							style={{ color: 'white' }}
						>
							{CST.BEETHOVEN.toUpperCase()}
						</a>
					</SCardTitle>
				}
				width={mobile ? '100%' : '420px'}
				margin={mobile ? '0' : '0 0 0 10px'}
				extra={
					<SCardExtraDivSolid>
						<Tooltip title={tooltioText}>
							<div>{states.state}</div>
							<img
								src={states.state === CST.CTD_TRADING ? successIcon : warningIcon}
							/>
						</Tooltip>
					</SCardExtraDivSolid>
				}
			>
				<SDivFlexCenter horizontal padding="0 10px">
					<SCardList noMargin>
						<div className="status-list-wrapper">
							<ul>
								<li className="block-title">
									{states.state !== CST.CTD_LOADING ? (
										states.state === CST.CTD_TRADING ? (
											<div className="last-reset-title">
												<span>{CST.TH_LAST_RESET[locale]}</span>
												<span className="last-reset-title-span">
													{states.resetPriceTime
														? moment(states.resetPriceTime).format(
																'YYYY-MM-DD HH:mm'
														)
														: CST.TH_LOADING[locale]}
												</span>
											</div>
										) : (
											CST.TH_RESET_PROGRESS[locale]
										)
									) : (
										CST.TH_LOADING[locale]
									)}
								</li>
								{states.state !== CST.CTD_LOADING ? (
									states.state === CST.CTD_TRADING ? (
										<li>
											<span className="title">ETH</span>
											<span className="content">
												{d3.formatPrefix(',.2', 1)(states.resetPrice) +
													' USD'}
											</span>
										</li>
									) : (
										<li className="no-bg">
											<SCardListProgressBar
												index={states.nextResetAddrIndex}
												total={states.totalUsers}
											>
												<div className="bar-bg">
													<div className="inner-bar" />
												</div>
												<div className="bar-text">
													{d3.format('.2%')(
														states.nextResetAddrIndex /
															(states.totalUsers || 1)
													)}
												</div>
											</SCardListProgressBar>
										</li>
									)
								) : (
									<li>
										<span className="title">{CST.TH_LOADING[locale]}</span>
									</li>
								)}
							</ul>
							<ul style={{ paddingTop: '10px' }}>
								<li className="block-title">{CST.TH_CONTRACT_STATES[locale]}</li>
								<li>
									<span className="title">{CST.TH_MATURITY[locale]}</span>
									<span className="content">
										{states.maturity
											? moment(states.maturity).format('YYYY-MM-DD HH:mm')
											: CST.TENOR_PPT}
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_PERIOD_LENGTH[locale]}</span>
									<span className="content">
										{states.period / 60000 + ' mins'}
									</span>
								</li>
								<li>
									<span className="title">
										{CST.TH_COUPON_PER_PERIOD[locale]}
									</span>
									<span className="content">
										{d3.format('.4%')(states.periodCoupon)}
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_UPPER_A[locale]}</span>
									<span className="content">
										{states.limitPeriodic
											? d3.format(',.4f')(states.limitPeriodic) + ' USD'
											: '-'}
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_UPPER_B[locale]}</span>
									<span className="content">
										{d3.format(',.4f')(states.limitUpper)} USD
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_LOWER_B[locale]}</span>
									<span className="content">
										{d3.format(',.4f')(states.limitLower)} USD
									</span>
								</li>
								<li>
									<span className="title">
										{CST.TH_LEVERAGE_FACTOR[locale] + ' (Alpha)'}
									</span>
									<span className="content">
										{d3.format(',.2f')(states.alpha)}
									</span>
								</li>
								<li>
									<span className="title">
										{CST.TH_CONVERSION_FACTOR[locale] + ' (Beta)'}
									</span>
									<span className="content">
										{d3.format(',.2f')(states.beta)}
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_ETH_BALANCE[locale]}</span>
									<span className="content">
										{d3.format(',.2f')(states.ethCollateral)}
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_A_SUPPLY[locale]}</span>
									<span className="content">
										{d3.format(',.2f')(states.totalSupplyA)}
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_B_SUPPLY[locale]}</span>
									<span className="content">
										{d3.format(',.2f')(states.totalSupplyB)}
									</span>
								</li>
								<li>
									<span className="title">{CST.TH_TOTAL_USERS[locale]}</span>
									<span className="content">{states.totalUsers}</span>
								</li>
							</ul>
						</div>
					</SCardList>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
