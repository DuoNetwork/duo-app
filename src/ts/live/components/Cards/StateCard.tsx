import { Tooltip } from 'antd';
import * as d3 from 'd3';
import moment from 'moment';
import * as React from 'react';
import successIcon from '../../../../images/stauts/success.svg';
import warningIcon from '../../../../images/stauts/warning.svg';
import * as CST from '../../common/constants';
import { ICustodianPrice, ICustodianStates } from '../../common/types';
import { SDivFlexCenter } from '../_styled';
import { SCard, SCardExtraDivSolid, SCardList, SCardListProgressBar, SCardTitle } from './_styled';

interface IProps {
	reset: ICustodianPrice;
	states: ICustodianStates;
}

export default class StateCard extends React.PureComponent<IProps> {
	public render() {
		const { states, reset } = this.props;
		const tooltioText =
			states.state === CST.CTD_TRADING
				? 'Trading state, operations are permitted during current state.'
				: states.state === CST.CTD_INCEPTION
					? 'Inception ongoing, please wait until the system is ready.'
					: 'Reset ongoing, operations are prohibited during current state.';
		return (
			<SCard
				title={
					<SCardTitle>
						<a
							className="tag-content"
							href={
								'https://' +
								(__KOVAN__ ? 'kovan.' : '') +
								'etherscan.io/address/' +
								(__KOVAN__ ? CST.CUSTODIAN_ADDR_KOVAN : CST.CUSTODIAN_ADDR_MAIN)
							}
							target="_blank"
							style={{ color: 'white' }}
						>
							{CST.TH_BEETHOVEN.toUpperCase()}
						</a>
					</SCardTitle>
				}
				width="420px"
				margin="0 0 0 10px"
				extra={
					<SCardExtraDivSolid>
						<Tooltip title={tooltioText}>
							<div>{states.state}</div>
							<img
								src={states.state === CST.CTD_TRADING ? successIcon : warningIcon}
							/>
						</Tooltip>
					</SCardExtraDivSolid>
				}
			>
				<SDivFlexCenter horizontal padding="0 10px">
					<SCardList noMargin>
						<div className="status-list-wrapper">
							<ul>
								<li className="block-title">
									{states.state !== CST.CTD_LOADING ? (
										states.state === CST.CTD_TRADING ? (
											<div className="last-reset-title">
												<span>Last Reset</span>
												<span className="last-reset-title-span">
													{reset.timestamp
														? moment(reset.timestamp).format(
																'YYYY-MM-DD HH:mm'
															)
														: 'Loading'}
												</span>
											</div>
										) : (
											'Reset Progress'
										)
									) : (
										CST.CTD_LOADING
									)}
								</li>
								{states.state !== CST.CTD_LOADING ? (
									states.state === CST.CTD_TRADING ? (
										<li>
											<span className="title">ETH</span>
											<span className="content">
												{d3.formatPrefix(',.2', 1)(reset.price) + ' USD'}
											</span>
										</li>
									) : (
										<li className="no-bg">
											<SCardListProgressBar
												index={states.nextResetAddrIndex}
												total={states.usersLength}
											>
												<div className="bar-bg">
													<div className="inner-bar" />
												</div>
												<div className="bar-text">
													{d3.format('.2%')(
														states.nextResetAddrIndex /
															(states.usersLength || 1)
													)}
												</div>
											</SCardListProgressBar>
										</li>
									)
								) : (
									<li>
										<span className="title">Loading</span>
									</li>
								)}
							</ul>
							<ul>
								<li className="block-title">Contract States</li>
								<li>
									<span className="title">Period Length</span>
									<span className="content">{states.period / 60 + ' mins'}</span>
								</li>
								<li>
									<span className="title">Coupon per Period</span>
									<span className="content">
										{d3.format('.4%')(states.periodCoupon)}
									</span>
								</li>
								<li>
									<span className="title">Upper Limit for Token A</span>
									<span className="content">
										{d3.format(',.4f')(states.limitPeriodic)} USD
									</span>
								</li>
								<li>
									<span className="title">Upper Limit for Token B</span>
									<span className="content">
										{d3.format(',.4f')(states.limitUpper)} USD
									</span>
								</li>
								<li>
									<span className="title">Lower Limit for Token B</span>
									<span className="content">
										{d3.format(',.4f')(states.limitLower)} USD
									</span>
								</li>
								<li>
									<span className="title">Leverage Factor (Alpha)</span>
									<span className="content">
										{d3.format(',.2f')(states.alpha)}
									</span>
								</li>
								<li>
									<span className="title">Conversion Factor (Beta)</span>
									<span className="content">
										{d3.format(',.2f')(states.beta)}
									</span>
								</li>
								<li>
									<span className="title">ETH:DUO Fee Ratio</span>
									<span className="content">1:{states.ethDuoFeeRatio}</span>
								</li>
								<li>
									<span className="title">DUO Received</span>
									<span className="content">
										{d3.format(',.2f')(states.duoBalance)}
									</span>
								</li>
								<li>
									<span className="title">Net ETH Balance</span>
									<span className="content">
										{d3.format(',.2f')(
											states.ethBalance - states.feeAccumulated
										)}
									</span>
								</li>
								<li>
									<span className="title">Token A Total Supply</span>
									<span className="content">
										{d3.format(',.2f')(states.totalSupplyA)}
									</span>
								</li>
								<li>
									<span className="title">Token B Total Supply</span>
									<span className="content">
										{d3.format(',.2f')(states.totalSupplyB)}
									</span>
								</li>
								<li>
									<span className="title">Total Users</span>
									<span className="content">{states.usersLength}</span>
								</li>
							</ul>
						</div>
					</SCardList>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
