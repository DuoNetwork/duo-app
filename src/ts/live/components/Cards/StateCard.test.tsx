import { shallow } from 'enzyme';
import * as React from 'react';
import StateCard from './StateCard';

describe('StateCard Test', () => {
	describe('User Login', () => {
		const reset = {
			address: '0x00476E55e02673B0E4D2B474071014D5a366Ed4E',
			price: 252.93877621546696,
			timestamp: 1534258801000
		};
		const states = {
			addrPoolLength: 6,
			adminCoolDown: 3600,
			alpha: 1,
			beta: 1,
			createCommRate: 0.01,
			duoBalance: 2482128.9762267205,
			ethBalance: 742.408092589115,
			ethDuoFeeRatio: 800,
			feeAccumulated: 87.06788278541666,
			iterationGasThreshold: 65000,
			lastAdminTime: 0,
			limitLower: 0.25,
			limitPeriodic: 1.035,
			limitUpper: 2,
			navA: 1.026299,
			navB: 0.5862017153842389,
			nextResetAddrIndex: 0,
			numOfPrices: 0,
			period: 3600,
			periodCoupon: 0.000017,
			preResetWaitingBlocks: 10,
			priceFeedTimeTol: 60,
			priceFeedTol: 0.01,
			priceTol: 0.05,
			priceUpdateCoolDown: 3000,
			redeemCommRate: 0.01,
			state: 'Trading',
			totalSupplyA: 82880.4753362674,
			totalSupplyB: 82880.4753362674,
			usersLength: 295
		};
		const mobile = true;
		const locale = 'EN';
		it('Test Snapshot', () => {
			const wrapper = shallow(
				<StateCard reset={reset} states={states} mobile={mobile} locale={locale} />
			);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
