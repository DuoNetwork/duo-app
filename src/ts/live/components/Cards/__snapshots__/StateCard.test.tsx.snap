// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateCard Test User Login Test Snapshot 1`] = `
<Styled(Card)
  extra={
    <styled.div>
      <Tooltip
        arrowPointAtCenter={false}
        autoAdjustOverflow={true}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="top"
        prefixCls="ant-tooltip"
        title="Trading state, operations are permitted during current state"
        transitionName="zoom-big-fast"
      >
        <div>
          Trading
        </div>
        <img
          src={null}
        />
      </Tooltip>
    </styled.div>
  }
  margin="0"
  title={
    <styled.div>
      <a
        className="tag-content"
        href="https://kovan.etherscan.io/address/0x0f80F055c7482b919183EcD06e0dd5FD7991D309"
        style={
          Object {
            "color": "white",
          }
        }
        target="_blank"
      >
        BEETHOVEN
      </a>
    </styled.div>
  }
  width="100%"
>
  <styled.div
    horizontal={true}
    padding="0 10px"
  >
    <styled.div
      noMargin={true}
    >
      <div
        className="status-list-wrapper"
      >
        <ul>
          <li
            className="block-title"
          >
            <div
              className="last-reset-title"
            >
              <span>
                Last Reset
              </span>
              <span
                className="last-reset-title-span"
              >
                2018-08-14 23:00
              </span>
            </div>
          </li>
          <li>
            <span
              className="title"
            >
              ETH
            </span>
            <span
              className="content"
            >
              252.94 USD
            </span>
          </li>
        </ul>
        <ul>
          <li
            className="block-title"
          >
            Contract States
          </li>
          <li>
            <span
              className="title"
            >
              Period Length
            </span>
            <span
              className="content"
            >
              60 mins
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Coupon per Period
            </span>
            <span
              className="content"
            >
              0.0017%
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Upper Limit for Token A
            </span>
            <span
              className="content"
            >
              1.0350
               USD
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Upper Limit for Token B
            </span>
            <span
              className="content"
            >
              2.0000
               USD
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Lower Limit for Token B
            </span>
            <span
              className="content"
            >
              0.2500
               USD
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Leverage Factor (Alpha)
            </span>
            <span
              className="content"
            >
              1.00
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Conversion Factor (Beta)
            </span>
            <span
              className="content"
            >
              1.00
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              ETH:DUO Fee Ratio
            </span>
            <span
              className="content"
            >
              1:
              800
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              DUO Received
            </span>
            <span
              className="content"
            >
              2,482,128.98
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Net ETH Balance
            </span>
            <span
              className="content"
            >
              655.34
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Token A Total Supply
            </span>
            <span
              className="content"
            >
              82,880.48
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Token B Total Supply
            </span>
            <span
              className="content"
            >
              82,880.48
            </span>
          </li>
          <li>
            <span
              className="title"
            >
              Total Users
            </span>
            <span
              className="content"
            >
              295
            </span>
          </li>
        </ul>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;
